### PR TITLE
feat: add arm64 support

### LINKED ISSUE #138: Support For ARM?
**Is your feature request related to a problem?/Why is this needed**
Can't install this on ARM nodes. The image repo does not seem to have anything ARM architecture related. 

**Describe the solution you'd like in detail**
Compile for ARM. Push ARM image to public registry. 

**Describe alternatives you've considered**
Compiling and building myself . . . probably not gonna work for me. 

**Additional context**
I am running a k3s cluster on a couple of Raspberry PIs. My router has really good Samba support with a USB 3.0 drive. I want to use this SMB server from the router to serve volumes for my cluster.

### PR DESCRIPTION
**Pull request by mayankshah1607**
Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #138

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: .github/workflows/image-push.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        uses: crazy-max/ghaction-docker-buildx@v3
+        with:
+          buildx-version: latest
+          qemu-version: latest
+      - name: Build and push ARM image
+        run: |
+          docker buildx build \
+          docker push $REPO:$TAG -t $REPO:$TAG \
+            --platform=linux/arm64,linux/amd64 \
+            --output="type=image" .
```
Comment (mayankshah1607) [inline .github/workflows/image-push.yaml:23]: @andyzhangx We need to set the registry details here. Could you share information regarding the registry / credentials? Thanks

Comment (andyzhangx) [reply .github/workflows/image-push.yaml:23]: `make smb-container` is here:
https://github.com/kubernetes-csi/csi-driver-smb/blob/e9c2dd3a6ad5d62670177a16cad9b37f4bdabf98/Makefile#L112-L121

I think you need to add one more command to build ARM64 container?

File: Makefile [hunk 2]
```diff
@@ -117,6 +117,7 @@ ifdef CI
 	make smb smb-windows
 	docker buildx build --no-cache --build-arg LDFLAGS=${LDFLAGS} -t $(IMAGE_TAG)-linux-amd64 -f ./pkg/smbplugin/Dockerfile --platform="linux/amd64" --push .
 	docker buildx build --no-cache --build-arg LDFLAGS=${LDFLAGS} -t $(IMAGE_TAG)-windows-1809-amd64 -f ./pkg/smbplugin/Windows.Dockerfile --platform="windows/amd64" --push .
+	docker buildx build --no-cache --build-arg LDFLAGS=${LDFLAGS} -t $(IMAGE_TAG)-linux-arm64 -f ./pkg/smbplugin/Dockerfile --platform="linux/arm64" --push .
```
Comment (mayankshah1607) [inline Makefile:120]: @andyzhangx I think this would do. But is `make smb-container` being called in the CI? I., the post image job?

Comment (andyzhangx) [reply Makefile:120]: let's wait for the test result, I have credentials to push image to Azure Container Registry.

Comment (andyzhangx) [reply Makefile:120]: I think it's better move L120 to L119

**Comment by andyzhangx** [comment]
there is error:
```
docker buildx build --no-cache --build-arg LDFLAGS="-X github.com/kubernetes-csi/csi-driver-smb/pkg/smb.driverVersion=e2e-25526bb608f502e89d3f6974ad9e94bce9ae1855 -X github.com/kubernetes-csi/csi-driver-smb/pkg/smb.gitCommit=25526bb608f502e89d3f6974ad9e94bce9ae1855 -X github.com/kubernetes-csi/csi-driver-smb/pkg/smb.buildDate=2020-10-30T14:41:35Z -s -w -extldflags '-static'" -t k8sprow.azurecr.io/smb-csi:e2e-25526bb608f502e89d3f6974ad9e94bce9ae1855-linux-arm64 -f ./pkg/smbplugin/Dockerfile --platform="linux/arm64" --push .
time="2020-10-30T14:44:42Z" level=warning msg="invalid non-bool value for BUILDX_NO_DEFAULT_LOAD: "
#1 [internal] load .dockerignore
#1 transferring context: 2B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 32B done
#2 DONE 0.0s

#3 [internal] load metadata for k8s.gcr.io/build-image/debian-base-amd64:v2...
#3 DONE 0.1s

#4 [internal] load metadata for docker.io/library/golang:1.13.10-alpine3.10
#4 DONE 1.4s
failed to solve: rpc error: code = Unknown desc = failed to load LLB: runtime execution on platform linux/arm64 not supported
make[2]: *** [Makefile:118: smb-container] Error 1
make[2]: Leaving directory '/home/prow/go/src/sigs.k8s.io/csi-driver-smb'
make[1]: *** [Makefile:74: e2e-bootstrap] Error 2
make[1]: Leaving directory '/home/prow/go/src/sigs.k8s.io/csi-driver-smb'
Failure [131.374 seconds]
```

not sure whether it's container image build error or Azure Container Registry not  supported issue, need to find out later.

**Comment by mayankshah1607** [comment]
Relevant - https://github.com/docker/buildx/issues/138
Do we have qemu installed in the machine running the job?

**Comment by kferrone** [comment]
Any luck? I've been following along

**Comment by mayankshah1607** [comment]
It looks like the [base image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) being used for the failing jobs does not have qemu installed, which is a requirement for building arm64 images (as per [this](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) comment) . 
Is it feasible modify the [base image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) and have qemu installed, so that buildx can successfully build arm64 images?

/Cc @andyzhangx @msau42

**Comment by andyzhangx** [comment]
> It looks like the [base image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) being used for the failing jobs does not have qemu installed, which is a requirement for building arm64 images (as per [this](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) comment) .
> Is it feasible modify the [base image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) and have qemu installed, so that buildx can successfully build arm64 images?
> 
> /Cc @andyzhangx @msau42

since it's just a testing image, I think it's ok, could you work on a PR to install qemu? thanks.

**Comment by andyzhangx** [comment]
> Any luck? I've been following along

@kferrone  I have already built the master branch image with ARM64 support with this PR, you could try now:
```yaml
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1163,
         "digest": "sha256:2d6616d944e5382175c824c3ff288723e91e736b6318dbd33f2a96cd6082371a",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1163,
         "digest": "sha256:6580abb9e16b557a10dd05a5ecc40f2b40c8d2874a50f6d0878dff118308a10c",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 950,
         "digest": "sha256:8f2bad9a5772a4f77774a707804a0f85f60b728723d8e56c770d01d5b500038e",
         "platform": {
            "architecture": "amd64",
            "os": "windows"
         }
      }
   ]
}
```

In the meantime, we will figure out how to make CI workflow works for ARM image, and

**Comment by andyzhangx** [comment]
> > It looks like the [base image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) being used for the failing jobs does not have qemu installed, which is a requirement for building arm64 images (as per [this](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) comment) .
> > Is it feasible modify the [base image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) and have qemu installed, so that buildx can successfully build arm64 images?
> > /Cc @andyzhangx @msau42
> 
> since it's just a testing image, I think it's ok, could you work on a PR to install qemu? thanks.

@mayankshah1607 I have pushed some changes in Makefile, it could build binary and docker image with ARM64 support now, following guide: https://www.docker.com/blog/multi-platform-docker-builds/

 - docker buildx setting to support ARM64
```console
export DOCKER_CLI_EXPERIMENTAL=enabled
docker run --privileged --rm docker/binfmt:a7996909642ee92942dcd6cff44b9b95f08dad64
```

**Comment by mayankshah1607** [comment]
@andyzhangx Tried adding the following to the Makefile - 
```
docker run --privileged --rm docker/binfmt:a7996909642ee92942dcd6cff44b9b95f08dad64
```
The error still remains. It doesn't look like the handlers are getting registered. Could it be because this is not meant to be run as DinD? In that case we could try to run ARM64 build and push separately on Github Actions

**Comment by andyzhangx** [comment]
> @andyzhangx Tried adding the following to the Makefile -
> 
> ```
> docker run --privileged --rm docker/binfmt:a7996909642ee92942dcd6cff44b9b95f08dad64
> ```
> 
> The error still remains. It doesn't look like the handlers are getting registered. Could it be because this is not meant to be run as DinD? In that case we could try to run ARM64 build and push separately on Github Actions

it could be, I think we can add a build flag in Makefile and then only run arm64 binary build and docker build image verification in GH action.

**Comment by mayankshah1607** [comment]
@andyzhangx It works now. I added the following line instead:
```
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```

**Comment by andyzhangx** [comment]
could you rebase this PR @mayankshah1607 thanks.

File: Makefile (hunk 2 above)
Comment (msau42) [reply Makefile:125]: Our Kubernetes ci pipeline supprts multiarch now including arm. But you need to setup the cloudbuild.yaml files like in the other csi repos

**Comment by mayankshah1607** [comment]
Following up with the [discussion on Slack](https://kubernetes.slack.com/archives/C09QZ4DQB/p1604941946039700?thread_ts=1604921217.035400&cid=C09QZ4DQB) regarding multi arch builds, we should move this to build in GCB, and not Prow directly.
> the prow jobs share a kernel and binfmt_misc is global / non-namespaced

Should we add the relevant build commands in a `cloudbuild.yaml` in this repo?

**Comment by andyzhangx** [comment]
> Following up with the [discussion on Slack](https://kubernetes.slack.com/archives/C09QZ4DQB/p1604941946039700?thread_ts=1604921217.035400&cid=C09QZ4DQB) regarding multi arch builds, we should move this to build in GCB, and not Prow directly.
> 
> > the prow jobs share a kernel and binfmt_misc is global / non-namespaced
> 
> Should we add the relevant build commands in a `cloudbuild.yaml` in this repo?

@mayankshah1607 since it already works, let's merge this PR first. I believe it would also benefit other CSI drivers too.

**Comment by andyzhangx** [comment]
and then continue with https://github.com/kubernetes-csi/csi-driver-smb/pull/28

**Comment by mayankshah1607** [comment]
Had some issues rebasing this branch after https://github.com/kubernetes-csi/csi-driver-smb/pull/141 was merged. Opening a new PR https://github.com/kubernetes-csi/csi-driver-smb/pull/144 and closing this one

**Comment by BenTheElder** [comment]
re: dind ... Prow is dind, but see: https://kubernetes.slack.com/archives/C09QZ4DQB/p1604941946039700?thread_ts=1604921217.035400&cid=C09QZ4DQB
GCB is not dind. (it is docker socket mounted to a container though, which is different, containers you create will be on the host adjacent to the build env)

**Pull request closed without merging**